components/script/dom/xrsession.rs

```diff
@@ -153,21 +155,15 @@ impl XRSession {
         self.mode != XRSessionMode::Inline
     }
 
-    fn setup_raf_loop(&self) {
-        assert!(
-            self.raf_sender.borrow().is_none(),
-            "RAF loop already set up"
-        );
+    fn setup_raf_loop(&self, frame_receiver: IpcReceiver<Frame>) {
         let this = Trusted::new(self);
         let global = self.global();
         let window = global.as_window();
         let (task_source, canceller) = window
             .task_manager()
             .dom_manipulation_task_source_with_canceller();
```

Comment:
To reduce the time between when we're signalled with a new frame and when we run the rAF callback, we should use a high-priority task source. The spec doesn't actually specify a particular ordering, so we should be able to get away with jumping the queue. For example, we could use a dedicated sender/receiver pair and use try_recv on it before calling select! in ScriptThread::handle_msgs.

Reply:
https://github.com/jdm/servo/tree/xr-frame-prioritize

Reply:
I think that's a good idea, and maybe it could be further improved by not going through the router to receive the frame. 

I think it could be done through:

1. Whenever a callback is added:
    - set a `awaiting_frame` flag on the script-thread.
    - send a `MainThreadScriptMsg::WakeUp` to ensure the main loop doesn't block on the select
2. After the [microtask checkpoint](https://github.com/jdm/servo/blob/871a805e124b11661adbee5dd63f473ed3661f20/components/script/script_thread.rs#L1589):
    - remove the `awaiting_frame` flag. 
    - block on the frame receiver, and 
    - immediately run the callbacks when receiving it. 
3. This would require also running 2 if `sequential` is empty(in the case that the only event is the `MainThreadScriptMsg::WakeUp`, since it will be filtered out by the task-queue). It also requires Xr to always send a message on the frame receiver to unblock the script-thread, even if no frame can be provided(in response to which you'd have to unset the flag on the script-thread). 

So that could be a way to block directly on a `IpcReceiver` inside `handle_msgs`, with the benefit being always running the frame callbacks immediately after you receive the frame, and not going through the router.  

Then I guess the script-thread could make the [`xr_frame_` pair](https://github.com/servo/servo/compare/master...jdm:xr-frame-prioritize#diff-1879ac6bd5d567e2aa43529d33474677R711) an ipc-channel where the sender would be cloned inside `request_new_xr_frame` and send to Xr?

Reply:
Ah ok I see you also need a way to call from the script-thread into the session, so instead of an `awaiting_frame` flag, you could have an `Option<SessionAwaitingFrame>` on the script-thread, which would be set by `request_new_xr_frame` and taken before blocking on the frame receiver from within `handle_msgs`. I assume you can have only one active Xr session per script-thread?

Reply:
Like https://github.com/gterzian/servo/tree/restructure_xr_frame_handling
If it seems like a good idea I could just open a PR as a follow-up.

Reply:
You can have multiple online sessions, but only one immersive 

We first need to figure out why this is breaking the tests. I bet it's because of a race between data getting sent and read from the mock device.

Reply:
I think the problem with the timeouts is that with the current changes, the call to [`render_animation_frame`](https://github.com/servo/servo/blob/1b5f0ed15a70aa0da88ddc5e67329474f206adc1/components/script/dom/xrsession.rs#L350) from script only occurs if the session is immersive. 

Previously, [a new frame would be requested](https://github.com/servo/servo/blob/106a0185f404c2ce5011272a9191eb71a731e855/components/script/dom/xrsession.rs#L364) at the end of running callbacks, regardless of whether the session was immersive.

Reply:
That was indeed the issue, was poking at this earlier. Fixed now.

Opening a followup PR would be great, but we should test if it affects FPS.